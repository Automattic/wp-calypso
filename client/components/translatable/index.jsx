/** @format */
/**
 * External dependencies
 */
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { noop } from 'lodash';
import Gridicon from 'gridicons';

/**
 * Internal dependencies
 */
import Popover from 'components/popover';
import FormFieldset from 'components/forms/form-fieldset';
import FormTextInput from 'components/forms/form-text-input';
import FormButton from 'components/forms/form-button';
import FormLabel from 'components/forms/form-label';


class Translatable extends Component {
	state = {
		showTooltip: false,
		translatedData: null,
	};

	timeout = null;

	hoverTriggerValue = 1000;

	refCallback = elem => ( this.elem = elem );

	togglePopover = event => {
		event.preventDefault();


		this.timeout = setTimeout( () => {
			this.setState( { showTooltip: true } );
			const { singular, context, plural } = this.props;

			/* temp xhr code for demonstration purposes */
			const xhr = new XMLHttpRequest();

			if ( 'withCredentials' in xhr ) {
				xhr.open( 'POST', 'https://translate.wordpress.com/api/translations/-query-by-originals', true );
				xhr.withCredentials = true;
				xhr.setRequestHeader( 'Content-type', 'application/x-www-form-urlencoded' );

				const data = {
					project: 'wpcom',
					locale_slug: 'it',
					original_strings: { singular, context, plural },
				};

				xhr.onreadystatechange = () => {
					if ( xhr.readyState > 3 && xhr.status === 200 ) {
						this.setState( { translatedData: JSON.parse( xhr.responseText ) } );
					};
				};
				const params = `project=wpcom&locale_slug=it&original_strings=${ encodeURIComponent( JSON.stringify( data ) ) }`;

				xhr.send( params );
			}
		}, this.hoverTriggerValue );

		if ( this.state.translatedData ) {
			return;
		}
	};

	resetTimer = () => {
		clearTimeout( this.timeout );
		// this.setState( { showTooltip: false } );
	}

	render() {
		const translatedStr = this.state.translatedData ? this.state.translatedData[ 0 ].translations[0][ 'translation_0' ] : '';
		// eslint-disable-next-line
		console.log( 'this.props.children', this.props.children );

		return (
			<data
				title={ 'do it' }
				className="translatable translatable__community-translator"
				onMouseEnter={ this.togglePopover }
				onMouseLeave={ this.resetTimer }
				ref={ this.refCallback }
				{ ...this.props }
			>
				{ this.props.children }

				{ this.state.showTooltip && <Popover
					isVisible={ this.state.showTooltip }
					context={ this.elem }
					onClose={ noop }
					position="bottom"
					className="translatable__popover popover"
				>
					<div className="translatable__popover-content">
						<h4>Original string</h4>
						<p>{ this.props.singular }</p>
						<p>Submit a new translation</p>
					</div>
				</Popover> }
			</data>
		);
	}
}

export default Translatable;
