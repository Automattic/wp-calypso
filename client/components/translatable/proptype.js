/** @format */

function translatableStringChecker( props, propName, componentName ) {
	componentName = componentName || 'ANONYMOUS';

	const value = props[ propName ];
	if ( value !== undefined && value !== null ) {
		if ( 'string' === typeof value ) {
			return null;
		}

		// Translator Jumpstart old-style
		if ( 'object' === typeof value && 'data' === value.type ) {
			return null;
		}

		// Translator Jumpstart after #21591
		if (
			'object' === typeof value &&
			'function' === typeof value.type &&
			( 'Translatable' === value.type.name ||
				// Accept HOC wrappings (e.g. `localize( Translatable )`)
				String( value.type.displayName ).match( /\(Translatable\)/ ) )
		) {
			return null;
		}

		return new Error(
			'Invalid value for Translatable string in `' +
				componentName +
				'`. Please pass a translate() call.'
		);
	}

	// assume all ok
	return null;
}

function createChainableTypeChecker( validate ) {
	function checkType( isRequired, props, propName, componentName, location ) {
		componentName = componentName || 'ANONYMOUS';
		if ( props[ propName ] === undefined ) {
			if ( isRequired ) {
				return new Error(
					'Required ' +
						location +
						' `' +
						propName +
						'` was not specified in ' +
						( '`' + componentName + '`.' )
				);
			}
			return null;
		}
		return validate( props, propName, componentName, location );
	}

	const chainedCheckType = checkType.bind( null, false );
	chainedCheckType.isRequired = checkType.bind( null, true );

	return chainedCheckType;
}

export default createChainableTypeChecker( translatableStringChecker );
