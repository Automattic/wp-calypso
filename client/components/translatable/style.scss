.translatable__community-translator {
	display: inline-block;
	position: relative;

	.translatable__globe {
		background-color: #dd3d36; // #ffba00 #4cb420
		display: inline-block;
		color: white;
		margin: 0 2px;
		cursor: alias;
	}
}

.translatable__popover-content {
	margin: 10px;
	text-align: left;
}


/*
.icons-globe {
	height: 12px ! important;
	width: 12px ! important;
	position: absolute ! important;
	fill: white ! important;
	top: auto ! important;
	left: 0 ! important;
	right: auto ! important;
	margin: 0 ! important;
	padding: 0 ! important;
}*/
