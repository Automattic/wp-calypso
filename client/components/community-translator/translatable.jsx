/**
 * External dependencies
 */
import React, { Component } from 'react';
import { isEmpty } from 'lodash';
import { localize } from 'i18n-calypso';
import classNames from 'classnames';
import Gridicon from 'calypso/components/gridicon';

/**
 * Internal dependencies
 */
import { Dialog, Button } from '@automattic/components';
import FormFieldset from 'calypso/components/forms/form-fieldset';
import TranslatableTextarea from './translatable-textarea';
import TranslatedSuccess from './translated-success';
import { getSingleTranslationData, getTranslationPermaLink, submitTranslation } from './utils.js';

/**
 * Module varialbles
 */
const TRANSLATION_FETCH_ERROR = 'translation-fetch-error';
const TRANSLATION_SUBMIT_ERROR = 'tranlsation-submit-error';

export class Translatable extends Component {
	state = {
		showDialog: false,
		originalData: {},
		formState: {},
	};

	hasDataLoaded() {
		return ! isEmpty( this.state.originalData ) || ! isEmpty( this.state.error );
	}

	handleTranslationChange = ( event ) => {
		const { name, value } = event.target;
		this.setState( {
			formState: {
				...this.state.formState,
				[ name ]: value,
			},
			submitting: false,
			submissionSuccess: false,
		} );
	};

	closeDialog = () => {
		this.setState( {
			showDialog: false,
			submissionSuccess: false,
			formState: this.state.submissionSuccess ? {} : this.state.formState,
			originalData: this.state.submissionSuccess ? {} : this.state.originalData,
		} );
	};

	openDialog = ( event ) => {
		event.preventDefault();

		this.setState( { showDialog: true } );

		const { singular, context, plural, locale } = this.props;

		if ( ! this.hasDataLoaded() ) {
			getSingleTranslationData( locale, { singular, context, plural } )
				.then( ( originalData ) =>
					this.setState( {
						originalData,
						translationUrl: getTranslationPermaLink( originalData.originalId, locale ),
						formState: {
							translatedSingular: originalData.translatedSingular,
							translatedPlural: originalData.translatedPlural,
						},
					} )
				)
				.catch( () => {
					this.setState( {
						error: TRANSLATION_FETCH_ERROR,
					} );
				} );
		}
	};

	submitForm = () => {
		this.setState( {
			submitting: true,
		} );
		submitTranslation( this.state.originalData.originalId, this.state.formState, this.props.locale )
			.then( ( originalData ) => {
				this.setState( {
					error: originalData.error,
					originalData,
					submitting: false,
					submissionSuccess: true,
				} );
			} )
			.catch( () => {
				this.setState( {
					error: TRANSLATION_SUBMIT_ERROR,
				} );
			} );
	};

	getDialogButtons = () => {
		const { translate } = this.props;

		const buttons = [
			<Button onClick={ this.closeDialog }>{ translate( 'Close', { textOnly: true } ) }</Button>,
		];

		if ( ! this.state.submissionSuccess ) {
			buttons.push(
				<Button
					primary
					onClick={ this.submitForm }
					disabled={
						this.state.originalData.translatedSingular ===
							this.state.formState.translatedSingular &&
						this.state.originalData.translatedPlural === this.state.formState.translatedPlural &&
						! this.state.submitting
					}
				>
					{ translate( 'Submit translation' ) }
				</Button>
			);
		}

		return buttons;
	};

	renderPlaceholder() {
		return (
			<div className="community-translator__string-container placeholder">
				<span className="community-translator__string-description" />
				<span />
			</div>
		);
	}

	getErrorMessage( errorType ) {
		const { translate } = this.props;

		switch ( errorType ) {
			case TRANSLATION_FETCH_ERROR:
				return translate( "Sorry, we couldn't find the translation for this string." );
			case TRANSLATION_SUBMIT_ERROR:
				return translate( "Sorry, we couldn't submit the translation for this string." );
			default:
				return translate( "Sorry, we've encountered an error." );
		}
	}

	renderTranslatableContent() {
		const { error, submissionSuccess, originalData, submitting, formState } = this.state;

		if ( error ) {
			return (
				<p className="community-translator__string-container">
					{ this.getErrorMessage( this.state.error ) }
				</p>
			);
		}

		if ( submissionSuccess ) {
			return <TranslatedSuccess translationUrl={ this.state.translationUrl } />;
		}

		return [
			originalData.comment && <p key="translationComment">{ originalData.comment }</p>,

			<TranslatableTextarea
				key="translatedSingular"
				originalString={ this.props.singular }
				title="Singular"
				fieldName="translatedSingular"
				onChange={ this.handleTranslationChange }
				disabled={ submitting }
				value={ formState.translatedSingular }
			/>,

			this.state.formState.translatedPlural && (
				<TranslatableTextarea
					key="translatedPlural"
					originalString={ this.props.plural }
					title="Plural"
					fieldName="translatedPlural"
					onChange={ this.handleTranslationChange }
					disabled={ submitting }
					value={ formState.translatedPlural }
				/>
			),
		];
	}

	renderDialogContent() {
		const { translate } = this.props;

		return (
			<div className="community-translator__dialog-content">
				<header className="community-translator__dialog-header">
					<h2>
						{ translate( 'Translate to %(localeName)s', {
							args: { localeName: this.props.locale.name },
						} ) }
					</h2>
					<nav>
						{ this.state.translationUrl && (
							<a
								target="_blank"
								rel="noopener noreferrer"
								title={ translate( 'Open this translation in translate.wordpress.com' ) }
								href={ this.state.translationUrl }
								className="community-translator__nav-link"
							>
								<Gridicon icon="external" size={ 12 } />
							</a>
						) }
						<a
							title={ translate( 'Settings' ) }
							href={ '/me/account' }
							className="community-translator__settings-link"
						>
							<Gridicon icon="cog" size={ 12 } onClick={ this.closeDialog } />
						</a>
						<Gridicon icon="help" className="community-translator__nav-link" size={ 12 } />
						<Gridicon
							icon="cross-circle"
							className="community-translator__close-icon"
							size={ 12 }
							onClick={ this.closeDialog }
						/>
					</nav>
				</header>
				<section className="community-translator__dialog-body">
					<FormFieldset>
						{ this.hasDataLoaded() ? this.renderTranslatableContent() : this.renderPlaceholder() }
					</FormFieldset>
				</section>
			</div>
		);
	}

	render() {
		const { untranslated, children } = this.props;

		const classes = classNames( 'translatable community-translator__element', {
			'is-untranslated': untranslated,
		} );

		return (
			<data className={ classes } onContextMenu={ this.openDialog }>
				{ children }

				{ this.state.showDialog && (
					<Dialog
						isVisible={ this.state.showDialog }
						buttons={ this.getDialogButtons() }
						additionalClassNames="community-translator__dialog"
					>
						{ this.renderDialogContent() }
					</Dialog>
				) }
			</data>
		);
	}
}

export default localize( Translatable );
