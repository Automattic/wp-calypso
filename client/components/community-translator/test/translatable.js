/**
 * @jest-environment jsdom
 */
/**
 * External dependencies
 */
import React from 'react';
import { shallow } from 'enzyme';
/**
 * Internal dependencies
 */

import { Translatable } from '../translatable';

const defaultProps = {
	singular: 'Original translation',
	locale: 'it',
	plural: 'Original translations',
};

describe( '<Translatable /> ', () => {
	test( 'should render', () => {
		const wrapper = shallow( <Translatable { ...defaultProps } /> );
		expect( wrapper ).toMatchSnapshot();
	} );
} );
