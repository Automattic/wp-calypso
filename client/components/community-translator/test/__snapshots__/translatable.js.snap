// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Translatable />  should render 1`] = `
<data
  className="translatable community-translator__element"
  onContextMenu={[Function]}
/>
`;
